src/Components/Components/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IAsyncDisposable
+    {
+        private int _itemsAbove;
+
+        private int _itemsVisible;
+
+        private ElementReference _topSpacer;
+
+        private ElementReference _bottomSpacer;
+
+        private IVirtualizationHelper? _virtualizationHelper;
+
+        [Inject]
+        private IVirtualizationService VirtualizationService { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the total number of items in the collection.
+        /// </summary>
+        protected abstract int ItemCount { get; }
+
+        /// <summary>
+        /// Gets the size (height) of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="RenderFragment"/>s representing a range of items.
+        /// </summary>
+        /// <param name="start">The start index of the range of items.</param>
+        /// <param name="count">The number of items in the range.</param>
+        /// <returns>The <see cref="RenderFragment"/>s representing the given range.</returns>
+        protected abstract IEnumerable<RenderFragment> GetItems(int start, int count);
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
```

Comment:
Suggestion: do this by overriding `SetParametersAsync` instead of `OnParametersSet`. This is because when developers override `OnParametersSet`, calling `base.OnParametersSet` is meant to be optional, and they might not bother to call it and then the check would not be enforced.

`SetParametersAsync` is a bit different in that it's considered a more low-level and advanced thing to override. We don't tell people to override it in most cases. In the rare cases where they do override it, they do have to call base.

I know that's a bit of a tenuous distinction but it's worked out fine so far in the other built-in components.